x/programs/runtime/runtime.go

```diff
@@ -5,178 +5,220 @@ package runtime
-	allocFnName   = "alloc"
-	deallocFnName = "dealloc"
-)
+var _ Runtime = &runtime{}
 
-func New(log logging.Logger, meter Meter, storage Storage) *runtime {
+// New returns a new wasm runtime.
+func New(log logging.Logger, cfg *Config, imports Imports) *runtime {
+	if imports == nil {
+		imports = make(Imports)
+	}
 	return &runtime{
+		imports: imports,
 		log:     log,
-		meter:   meter,
-		storage: storage,
+		cfg:     cfg,
 	}
 }
 
 type runtime struct {
+	cfg   *Config
+	inst  *wasmtime.Instance
+	store *wasmtime.Store
+	mod   *wasmtime.Module
+	exp   WasmtimeExportClient
+	meter Meter
+
+	once     sync.Once
 	cancelFn context.CancelFunc
-	engine   wazero.Runtime
-	mod      api.Module
-	meter    Meter
-	storage  Storage
-	// functions exported by this runtime
-	mu state.Mutable
-
-	closed bool
-	log    logging.Logger
-}
 
-func (r *runtime) Create(ctx context.Context, programBytes []byte) (uint64, error) {
-	err := r.Initialize(ctx, programBytes)
-	if err != nil {
-		return 0, err
-	}
-	// get programId
-	programID := initProgramStorage()
+	imports Imports
 
-	// call initialize if it exists
-	result, err := r.Call(ctx, "init", uint64(programID))
-	if err != nil {
-		if !errors.Is(err, ErrMissingExportedFunction) {
-			return 0, err
-		}
-	} else {
-		// check boolean result from init
-		if result[0] == 0 {
-			return 0, fmt.Errorf("failed to initialize program")
-		}
-	}
-	return uint64(programID), nil
+	log logging.Logger
 }
 
-func (r *runtime) Initialize(ctx context.Context, programBytes []byte) error {
+func (r *runtime) Initialize(ctx context.Context, programBytes []byte) (err error) {
 	ctx, r.cancelFn = context.WithCancel(ctx)
+	go func(ctx context.Context) {
+		<-ctx.Done()
+		r.log.Debug("runtime context done, stopping...")
+		// send immediate interrupt to engine
+		r.Stop()
+	}(ctx)
+
+	r.store = wasmtime.NewStore(wasmtime.NewEngineWithConfig(r.cfg.engine))
+	r.store.Limiter(
+		r.cfg.limitMaxMemory,
+		r.cfg.limitMaxTableElements,
+		r.cfg.limitMaxInstances,
+		r.cfg.limitMaxTables,
+		r.cfg.limitMaxMemories,
+	)
+
+	// set initial epoch deadline
+	r.store.SetEpochDeadline(1)
+
+	switch r.cfg.compileStrategy {
+	case PrecompiledWasm:
+		// Note: that to deserialize successfully the bytes provided must have been
+		// produced with an `Engine` that has the same compilation options as the
+		// provided engine, and from the same version of this library.
+		//
+		// A precompile is not something we would store on chain.
+		// Instead we would prefetch programs and precompile them.
+		r.mod, err = wasmtime.NewModuleDeserialize(r.store.Engine, programBytes)
+		if err != nil {
+			return err
+		}
+	case CompileWasm:
+		r.mod, err = wasmtime.NewModule(r.store.Engine, programBytes)
+		if err != nil {
+			return err
+		}
+	default:
+		return fmt.Errorf("unsupported compile strategy: %v", r.cfg.compileStrategy)
+	}
 
-	r.engine = wazero.NewRuntimeWithConfig(ctx, wazero.NewRuntimeConfigInterpreter())
+	// TODO: remove wasi host functions.
```

Comment:
Can you add this as an explicit TODO on the PR?

Reply:
done